### ISSUE TITLE
Proposal: first-class support for stack allocation

### ISSUE THREAD

**Issue by julien-h** [scala-native/scala-native#115]
This issue is based on the snippet shown in the readme, and concerns the syntax of stack allocation. 

For obvious performance reasons, stack allocation is the preferred way to allocate objects and structs local to a function. It is the default mode in C++ [and Rust](https://doc.rust-lang.org/book/the-stack-and-the-heap.html#which-to-use).

Hence, I expect most of the `val` and `var` in native-scala to be allocated on the stack, reserving GC for non-local objects.

However, the current way to use the stack is too verbose to be realistic. 

```
val vec = stackalloc[Vec] // pointer to stack allocation
!vec = new Vec(1, 2, 3)   // store value to stack
```
- We can expect that programmers will hesitate to allocate on the stack because of the associated syntax cost. 
- C++ (and Rust?) programmers might dismiss scala-native because of it. They are well too acustomed to the burden of pointer dereferencing and know too well the benefit of stack allocation to give it up. 
- And those who can't bear GC-ing something that should have been on the stack will hold the language responsible for their [RSI](https://en.wikipedia.org/wiki/Repetitive_strain_injury) (reaching for ! at every line is painful).

I propose one of the following syntax for stack allocation:

```
sval vec = new Vec(1, 2, 3) 
vec updated (2, 4)
```

or:

```
stack val vec = new Vec(1, 2, 3)
vec updated (2, 4)
```

or:

```
val& vec = new Vec(1, 2, 3)
vec updated (2, 4)
```

In any case, `vec` should be usable as a normal val reference, without the syntactic cost of dereferencing a pointer.

**Comment by sjrd** [comment]
(@densh should confirm, but I'm confident I can answer this:) Changing the syntax or type checking of Scala is impossible for practical reasons (not only is it hard to change that in the compiler, but it also prevents tools such as sbt or IDEs from understanding Scala Native), and therefore out-of-scope for this project.

Anyway, I think your concern is not founded. `stackalloc` is only needed for objects that you want allocated on stack but that still need to be passed (by reference) to other functions. Objects that are actually local to a method can be stack-allocated by an optimizer even if they are declared as `val vec = new Vec`. Therefore, I do not think your premise that most vals and vars will need `stackalloc` is valid.

Additionally, I wouldn't put too much weight on the appeal-to-C++-devs right now. From experience, I can tell you it took Scala.js 2.5 years of convincing *Scala* developers before we started advocating it to JavaScript developers. You cannot convince "foreign" devs if your language doesn't have a reasonable eco-system to support its credibility. And you cannot build that eco-system without first having Scala developers interested.

**Comment by julien-h** [comment]
Hi sjrd,

> stackalloc is only needed for objects that you want allocated on stack but that still need to be passed (by reference) to other functions. 

Yeah well, most local objects are created and then **passed to helper functions**. There is only a very restricted set of things you can do without helper functions (unless you write 200 lines functions, of course...).


> Additionally, I wouldn't put too much weight on the appeal-to-C++-devs right now. 

Agreed but the "syntax" should be fixed at the earliest possible stage to avoid the weight of (bad) habits.



> Changing the syntax or type checking of Scala is impossible for practical reasons 

How about : 

    @stack
    val vec = new Vec(1, 2, 3)

The thing with stack pointer is that they are not real pointers.
* You can't free it and reallocate it like you would with normal heap pointers. 
* Furthermore heap pointers can be returned by the function, while stack pointers can't. 
* Stack pointers won't outlive the function, so why are they pointers at all ?

**Comment by sjrd** [comment]
> Yeah well, most local objects are created and then passed to helper functions.

I would have to seriously think about it, but I have a draft of an optimizer algorithm that would still allow that to work. Basically, you traverse the call graph from escape points (e.g., when a reference is stored on the heap) backwards to allocation points to mark those allocation points as "need heap allocation". After this pass, unmarked allocations can all be optimized on stack.

**Comment by DarkDimius** [comment]
@julien-h @sjrd Escape analysis is one of the most simple analysis to implement and (almost) every dataflow algorithm is going to have enough information to subsume escape analysis. 
It's quite likely that I'll need a dataflow analysis quite soon in https://github.com/dotty-linker/dotty and building escape analysis on top of it is going to be trivial.

**Comment by densh** [comment]
Thanks for bring this topic up, @julien-h. I agree that familiarity is an important aspect and we would want to make Scala Native as unsurprising to existing developers as possible.

At the same time, as @sjrd pointed out, any syntax changes are out of scope. We're still Scala at the core, and don't plan on making any drastic changes to the language (i.e. no changes to syntax, type system and semantics of normal Scala features.) We can only *conservatively* augment existing language with additional things we need for interop with native code and for lower-level programming in general.

## State of things

Our current solution to stack allocation is two-fold:

* First of all we have normals `var`-s. Those are allocated on stack by default and are implicitly dereferenced for reads and writes. Because the stack pointer is not exposed in any way and never escapes the function scope, those are *guaranteed* to be optimised away by LLVM to remove the stack allocation altogether and replace it with control-flow dependent SSA equivalent. 

  **Note**: `val`-s are never stack-allocated, they map to SSA assignments directly and thus are completely zero-cost.

* We also have `def stackalloc[T]: Ptr[T]` as a means to do malloc-style allocations on stack. Those return an explicit pointer to the stack allocation. As @julien-h noticed, primary use case for this one is to support escaping of stack-allocated memory to helper functions.

C++'s solution to the same problem is quite a bit different:

* All variables are stack-allocated and are implicitly deferenced on accesses. You can trivially access a pointer to stack-allocated memory using `&` operator. This means that there is no guarantee that those will be optimised away in SSA.

* Additionally you have pass-by-reference semantics via `&` that lets one pass implicitly-derefenceable pointers across function boundaries without exposing them as true pointers. 

Another interesting data point is C#'s `ref` (see https://github.com/dotnet/roslyn/issues/118):

* Variables do not expose stack pointer by default, are implicitly dereferenced for reads/writes.

* There are special `ref` variables/parameters/returns that lets one pass implicit stack pointer across function boundaries. This is quite similar to C++'s by-reference passing but quite a bit more explicit and with extra safety guarantees.

## Moving forward

First of all, it's important to note that we *can not* adopt C++-style approach to variables as this would break base Scala semantics by adding support for mutations to vars that are outside current function scope. 

The two possible options that we have right now are:

* C#-style approach is quite compatible with Scala as `ref` variables did not exist before, and it would be clear that some more-than-normal-Scala magic is going thanks to explicit annotations. Those can be supported syntactically through explicit `@ref` annotations, similar to the `@stack` proposed above. Main issue here is implementation complexity and it's not clear if it carries its weight over having less syntactically convenient equivalent that we have now.

* Another possible solution to the problem would be to improve usability of pointer typesto make them easier to use by adding support for direct field accesses and method calls straight on the pointer type. This is way easier to implement and support compared to `@ref`-s.

**Comment by densh** [comment]
@sjrd @DarkDimius Escape analysis is easy as long as it's within a single function, it's quite hard if you want to support use cases like this:

```
@struct class Value(...)

def helper2(@ref value: Value) = ...
def helper1(@ref value: Value) = { ...; helper2(value); ... }
def entry = {
  @ref var x: Value = ...
  helper(1)
}
```

AFAIK HotSpot won't be able to do this unless both helpers are inlined. C++'s and C#'s pass-by-reference can do it even if they are not.

**Comment by julien-h** [comment]
So we won't need stack pointers at all then ?

**Comment by DarkDimius** [comment]
@densh I've been recently discussing with @olhotak a case where we would need a quite simple whole-program dataflow analysis. It is going to be simple as it would base on a precise call-graph. 

We have some research projects in mind that would require implementing such an analysis and the covered cases should include the case you've demonstrated(unless some of .... makes `x` escape to heap)

**Comment by julien-h** [comment]
@densh I think that improving usability on the pointer type is a good compromise since it does not impose a syntactic cost on the programmer and can also be applied to heap pointers (win/win).

I still need time to document and think about the `ref` alternative.

Are you in lausanne next Wednesday on the occasion of the LAMP Open House at EPFL ?

**Comment by densh** [comment]
@DarkDimius Looking forward to that. 👍 

@julien-h Yes, I'll be at the LAMP Open House.

**Comment by frgomes** [comment]
I vote for ``@stack`` annotation, which is clear and explicit. The compilation should fail in case the compiler determines that such intent expressed by the developer cannot be satisfied.

**Comment by densh** [comment]
We've reduced the scope of what structs are going to support in 0.1, in particular they are not first-class anymore and require explicit stackalloc at all times. The main motivation is to have simple solution now and improve it in the future. [A bit more details are available in the docs](http://scala-native.readthedocs.io/en/latest/user/interop.html#memory-layout-types).
